pkg/utils/gwlog/metadata.go

```diff
@@ -0,0 +1,86 @@
+package gwlog
+
+import (
+	"context"
+	"github.com/google/uuid"
+)
+
+type key string
+
+const metadata key = "metadata"
+
+const traceID string = "trace_id"
+
+type metadataValue struct {
+	m map[string]string
+}
+
+func (mv *metadataValue) get(key string) string {
+	return mv.m[key]
+}
+
+func (mv *metadataValue) set(key, val string) {
+	mv.m[key] = val
+}
+
+func newMetadata() *metadataValue {
+	return &metadataValue{
+		m: make(map[string]string),
+	}
+}
+
+func NewTrace(ctx context.Context) context.Context {
+	currID := uuid.New()
+
+	newCtx := context.WithValue(ctx, metadata, newMetadata())
+	AddMetadata(newCtx, traceID, currID.String())
+
+	return newCtx
+}
+
+func AddMetadata(ctx context.Context, key, value string) {
+	if ctx.Value(metadata) != nil {
+		ctx.Value(metadata).(*metadataValue).set(key, value)
+	}
```

Comment:
ack, I think doing nothing is fine but I might want to make it explicit

there's a conversation around if we want to add metadata even when a new trace is not started. I will follow up on that in a separate comment.

Reply:
I looked at this again and I think that the behavior here is correct. I imagine the developer would confirm that their metadata was correctly added in a particular log line.

Alternative is to add a log line or exit here to denote that the developer did something unintended, but seems rather aggressive.

Reply:
Thanks for having a look. I suppose most of the unit tests would also be sending through a placeholder context so it would get pretty chatty if we warned here.